### PR TITLE
add CoupledMaternGP to contrib.timeseries

### PR DESCRIPTION
**Pull request by martinjankowiak**
addresses #2090

### CODE REVIEW

File: pyro/contrib/timeseries/gp.py [hunk 1]
```diff
@@ -13,15 +14,14 @@ class IndependentMaternGP(TimeSeriesModel):
     with a univariate Gaussian Process with a Matern kernel. The targets are assumed
     to be evenly spaced in time. Training and inference are logarithmic in the length
     of the time series T.
+
```
Comment (martinjankowiak) [inline pyro/contrib/timeseries/gp.py]: moved this

File: pyro/ops/tensor_utils.py [hunk 2]
```diff
@@ -3,13 +3,14 @@
 
 def block_diag(m):
     """
-    Takes a 3-dimensional tensor of shape (B, M, N) and returns a block diagonal tensor
-    of shape (B x M, B x N).
+    Takes a of shape (..., B, M, N) and returns a block diagonal tensor
```
Comment (martinjankowiak) [inline pyro/ops/tensor_utils.py]: generalized

File: tests/contrib/timeseries/test_gp.py [hunk 3]
```diff
@@ -2,51 +2,73 @@
...
+def test_independent_matern_gp(model, nu, obs_dim, T):
     torch.set_default_tensor_type('torch.DoubleTensor')
     dt = 0.1 + torch.rand(1).item()
 
-    gp = IndependentMaternGP(nu=nu, obs_dim=obs_dim, dt=dt,
-                             log_length_scale_init=torch.randn(obs_dim),
-                             log_kernel_scale_init=torch.randn(obs_dim),
-                             log_obs_noise_scale_init=torch.randn(obs_dim))
+    if model == 'lcmgp':
+        if obs_dim == 1:
```
Comment (martinjankowiak) [inline tests/contrib/timeseries/test_gp.py]: lcmgp expects obs_dim > 1

File: tests/contrib/timeseries/test_gp.py [hunk 4]
```diff
@@ -2,51 +2,73 @@
...
+    if model == 'imgp':
+        assert gp_log_prob.shape == (obs_dim,)
+    else:
+        assert gp_log_prob.dim() == 0
 
-    for dim in range(obs_dim):
-        lengthscale = gp.kernel.log_length_scale.exp()[dim]
-        variance = (2.0 * gp.kernel.log_kernel_scale).exp()[dim]
-        obs_noise = (2.0 * gp.log_obs_noise_scale).exp()[dim]
+    if model == 'imgp':
```
Comment (martinjankowiak) [inline tests/contrib/timeseries/test_gp.py]: only imgp admits direct comparison to univariate gps via mvn

File: pyro/contrib/timeseries/gp.py [hunk 5]
```diff
@@ -86,11 +92,11 @@ def _forecast(self, dts, filtering_state, include_observation_noise=True):
...
         dts = dts.unsqueeze(-1).unsqueeze(-1).unsqueeze(-1)
-        trans_mat = self.kernel.transition_matrix(dts)
-        predicted_mean = torch.matmul(filtering_state.loc.unsqueeze(-2), trans_mat).squeeze(-2)[..., 0]
-        predicted_function_covar = torch.matmul(trans_mat.transpose(-1, -2),
+        trans_matrix, process_covar = self.kernel.transition_matrix_and_covariance(dt=dts)
+        predicted_mean = torch.matmul(filtering_state.loc.unsqueeze(-2), trans_matrix).squeeze(-2)[..., 0]
+        predicted_function_covar = torch.matmul(trans_matrix.transpose(-1, -2),
                                                 torch.matmul(filtering_state.covariance_matrix,
-                                                trans_mat))[..., 0, 0]
+                                                trans_matrix))[..., 0, 0] + process_covar[..., 0, 0]
```
Comment (martinjankowiak) [inline pyro/contrib/timeseries/gp.py]: add dropped term

File: pyro/contrib/timeseries/gp.py [hunk 6]
```diff
@@ -111,3 +117,142 @@ def forecast(self, targets, dts):
...
+        trans_mat = block_diag(trans_mat)  # S x full_state_dim x full_state_dim
+        process_covar = block_diag(process_covar)  # S x full_state_dim x full_state_dim
+        obs_matrix = self._get_obs_matrix()  # full_state_dim x obs_dim
+        trans_obs = torch.matmul(trans_mat, obs_matrix)  # S x full_state_dim x obs_dim
+        predicted_mean = torch.matmul(filtering_state.loc.unsqueeze(-2), trans_obs).squeeze(-2)
+        predicted_function_covar = torch.matmul(trans_obs.transpose(-1, -2),
+                                                torch.matmul(filtering_state.covariance_matrix,
+                                                trans_obs))
+        predicted_function_covar = predicted_function_covar + \
+            torch.matmul(obs_matrix.transpose(-1, -2), torch.matmul(process_covar, obs_matrix))
```
Comment (martinjankowiak) [inline pyro/contrib/timeseries/gp.py]: add dropped term

File: tests/contrib/timeseries/test_gp.py [hunk 7]
```diff
@@ -2,51 +2,74 @@
...
+            assert_equal(mvn_log_prob, gp_log_prob[dim])
 
-    for S in [1, 3]:
-        dts = torch.rand(S)
+    for S in [1, 5]:
+        dts = torch.rand(S).cumsum(dim=-1)
         predictive = gp.forecast(targets, dts)
         assert predictive.loc.shape == (S, obs_dim)
-        assert predictive.scale.shape == (S, obs_dim)
+        if model == 'imgp':
```
Comment (martinjankowiak) [inline tests/contrib/timeseries/test_gp.py]: monotonicity tests

File: tests/contrib/timeseries/test_gp.py [hunk 8]
```diff
@@ -2,51 +2,74 @@
...
+                dets = predictive.covariance_matrix.det()
+                delta = dets[1:S] - dets[0:S-1]
+                assert (delta > 0.0).sum() == (S - 1)
 
     # the distant future
     dts = torch.tensor([500.0])
     predictive = gp.forecast(targets, dts)
     # assert mean reverting
     assert_equal(predictive.loc, torch.zeros(1, obs_dim))
-    # assert large time covariance
```
Comment (martinjankowiak) [inline tests/contrib/timeseries/test_gp.py]: remove erroneous assertion

File: pyro/contrib/timeseries/gp.py (hunk 5 above)
Comment (fritzo) [reply pyro/contrib/timeseries/gp.py]: 1. Is this covered by a test?

2. It seems wasteful to compute matmuls and then index. Is it worth optimizing, say as
```py
torch.einsum("...a,...b,...ab->...",
             trans_matrix[..., 0], trans_matrix[..., 0],
             filtering_state.covariance_matrix)
```

File: pyro/contrib/timeseries/gp.py [hunk 9]
```diff
@@ -111,3 +117,142 @@ def forecast(self, targets, dts):
...
+
+
+class LinearlyCoupledMaternGP(TimeSeriesModel):
+    """
+    A time series model in which each output dimension is modeled as a linear combination
+    of shared univariate Gaussian Processes with Matern kernels.
+
+    In more detail, the generative process is:
+
+        :math:`y_i(t) = \\sum_j A_{ij} f_j(t) + \\epsilon_i`
```
Comment (fritzo) [inline pyro/contrib/timeseries/gp.py]: Could you mention that `f_j` is typically lower dimensional than `y_i`? I think that could help understand the motivation.

File: pyro/contrib/timeseries/gp.py [hunk 10]
```diff
@@ -111,3 +117,142 @@ def forecast(self, targets, dts):
...
+        self.full_state_dim = num_gps * self.kernel.state_dim
+
+        self.log_obs_noise_scale = nn.Parameter(log_obs_noise_scale_init)
+        self.A = nn.Parameter(0.3 * torch.randn(self.num_gps, self.obs_dim))
+
+    def _get_obs_matrix(self):
+        # (num_gps, obs_dim) => (state_dim * num_gps, obs_dim)
+        return self.A.repeat_interleave(self.kernel.state_dim, dim=0) * \
+            torch.tensor([1.0] + [0.0] * (self.kernel.state_dim - 1),
+                         dtype=self.A.dtype, device=self.A.device).repeat(self.num_gps).unsqueeze(-1)
```
Comment (fritzo) [inline pyro/contrib/timeseries/gp.py]: FYI you can also use the `.new_tensor()` method, e.g. this is equivalent:
```py
self.A.new_tensor([1.0] + [0.0] * (self.kernel.state_dim - 1)
                  ).repeat(self.num_gps).unsqueeze(-1)
```

File: tests/contrib/timeseries/test_gp.py (hunk 3 above)
Comment (fritzo) [reply tests/contrib/timeseries/test_gp.py]: nit: could you either `pytest.skipif(obs_dim == 1)` or winnow the parameters as
```py
@pytest.mark.parametrize('model,obs_dim', [('lcmgp', 3),
                                           ('imgp', 1),
                                           ('imgp', 3)])
```

File: examples/contrib/timeseries/gp_models.py [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+    # standardize data
+    data_mean = data[0:T_train, :].mean(0)
+    data -= data_mean
+    data_std = data[0:T_train, :].std(0)
+    data /= data_std
+
+    torch.manual_seed(args.seed)
+
+    # set up model
+    if args.model == "imgp":
```
Comment (martinjankowiak) [inline examples/contrib/timeseries/gp_models.py]: this was renamed but is basically the same except it now supports 2 models

File: examples/contrib/timeseries/gp_models.py [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+            # note that each call to forecast() conditions on more data than the previous call
+            dts = torch.tensor([1.0]).double()
+            pred_dist = gp.forecast(data[0:T_train + t, :], dts)
+            onestep_means[t, :] = pred_dist.loc.data.numpy()
+            if args.model == "imgp":
+                onestep_stds[t, :] = pred_dist.scale.data.numpy()
+            elif args.model == "lcmgp":
+                onestep_stds[t, :] = pred_dist.covariance_matrix.diagonal(dim1=-1, dim2=-2).data.numpy()
+
+        # do (non-rolling) multi-step forecasting
```
Comment (martinjankowiak) [inline examples/contrib/timeseries/gp_models.py]: we also do multi-step forecasting

File: pyro/contrib/timeseries/gp.py (hunk 5 above)
Comment (martinjankowiak) [reply pyro/contrib/timeseries/gp.py]: 1. this is partially tested by the "assert monotonic increase of predictive noise" test

Comment (martinjankowiak) [reply pyro/contrib/timeseries/gp.py]: ok let me try the einsum

File: pyro/contrib/timeseries/gp.py (hunk 9 above)
Comment (martinjankowiak) [reply pyro/contrib/timeseries/gp.py]: i mention this in the `num_gps` docstring

File: pyro/contrib/timeseries/gp.py (hunk 10 above)
Comment (martinjankowiak) [reply pyro/contrib/timeseries/gp.py]: done

File: tests/contrib/timeseries/test_gp.py (hunk 3 above)
Comment (martinjankowiak) [reply tests/contrib/timeseries/test_gp.py]: done

**Comment by martinjankowiak** [comment]
@fritzo address comments. i'm not sure how best to convince oneself of correctness here. i guess the monotonicity tests and the fact that various things remain PSD in addition to reasonable looking plots and reasonable LLs/RMSEs on a variety of datasets. i can't think of additional tests that would be useful

File: pyro/contrib/timeseries/gp.py [hunk 13]
```diff
@@ -86,11 +86,11 @@ def _forecast(self, dts, filtering_state, include_observation_noise=True):
...
-        trans_mat = self.kernel.transition_matrix(dts)
-        predicted_mean = torch.matmul(filtering_state.loc.unsqueeze(-2), trans_mat).squeeze(-2)[..., 0]
-        predicted_function_covar = torch.matmul(trans_mat.transpose(-1, -2),
-                                                torch.matmul(filtering_state.covariance_matrix,
-                                                trans_mat))[..., 0, 0]
+        trans_matrix, process_covar = self.kernel.transition_matrix_and_covariance(dt=dts)
+        predicted_mean = torch.matmul(filtering_state.loc.unsqueeze(-2), trans_matrix).squeeze(-2)[..., 0]
+        predicted_function_covar = torch.matmul(trans_matrix.transpose(-1, -2), torch.matmul(
+                                                filtering_state.covariance_matrix, trans_matrix))[..., 0, 0] + \
+            process_covar[..., 0, 0]
```
Comment (martinjankowiak) [inline pyro/contrib/timeseries/gp.py]: reverted einsum since it doesn't support broadcasting, which is a headache

**Comment by fritzo** [comment]
Re: tests. Let's try to add some stronger tests before the 1.0 release. Maybe these can be shared across multiple time series models. Something to think about.

**Pull request merged**